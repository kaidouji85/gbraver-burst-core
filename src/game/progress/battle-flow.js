// @flow

import type {Battle, BattleResult, Effect, GameState, PlayerCommand, PlayerState, TryReflect} from "../..";
import {gameFlow} from "./game-flow";
import {batteryDeclaration} from "../../effect/battery-declaration";
import {battle} from "../../effect/battle";
import {gameEndJudging} from "../end-judging";
import {gameEnd} from "../../effect/game-end";
import {turnChange} from "../../effect/turn-change";
import {inputCommand} from "../../effect/input-command";
import {reflect} from "../../effect/reflect";
import type {HistoryUpdate} from "./game-flow";

/**
 * 戦闘のフロー
 *
 * @param lastState 最後の状態
 * @param commands コマンド
 * @return 更新されたゲームの状態
 */
export function battleFlow(lastState: GameState, commands: PlayerCommand[]): GameState[] {
  return gameFlow(lastState, [
    state => [batteryDeclaration(state, commands)],
    state => [battle(state, commands)],
    state => {
      if (state.effect.name === 'Battle') {
        const battle: Battle = state.effect;
        return [
            ...(canReflectFlow(battle.result) ? reflectFlow(state) : [])
        ];
      }
      return [];
    },
    state => {
      const endJudge = gameEndJudging(state);
      if (endJudge.type !== 'GameContinue') {
        return [gameEnd(state, endJudge)];
      }

      return gameFlow(state, [
        v => [turnChange(v)],
        v => [inputCommand(v)]
      ]);
    }
  ]);
}

/**
 * ダメージ反射フローを実行できるか否かを判定する
 *
 * @param battle 戦闘結果
 * @return 判定結果、trueでダメージ反射フローを行う
 */
export function canReflectFlow(result: BattleResult): boolean {
  return result.name === 'NormalHit'
      || result.name === 'Guard'
      || result.name === 'CriticalHit';
}

/**
 * ダメージ反射のフロー
 * 本フローは戦闘直後に呼ばれる想定である
 *
 * @param lastState 最新状態
 * @return 更新結果
 */
export function reflectFlow(lastState: GameState): GameState[] {
  const attacker = lastState.players.find(v => v.playerId === lastState.activePlayerId);
  const defender = lastState.players.find(v => v.playerId !== lastState.activePlayerId);
  if (!attacker || !defender) {
    return [];
  }

  const attackerState: PlayerState = attacker;
  const defenderState: PlayerState = defender;
  const historyUpdates: HistoryUpdate[] = defenderState.armdozer.effects
    .filter(v => v.type === 'TryReflect')
    .map(v => {
      const tryReflect: TryReflect = (v: TryReflect);
      return (state: GameState): GameState[] => [reflect(state, attackerState.playerId, tryReflect.damage, tryReflect.effect)];
    });

  return gameFlow(lastState, historyUpdates);
}