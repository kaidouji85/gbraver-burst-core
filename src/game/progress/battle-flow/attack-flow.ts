import type { BatteryCommand } from "../../../command/battery";
import { batteryDeclaration } from "../../../effect/battery-declaration";
import { battle } from "../../../effect/battle";
import type { GameState } from "../../../state/game-state";
import type { PlayerCommandX } from "../../command/player-command";
import { startGameFlow } from "../../game-flow";

/**
 * プレイヤー攻撃フロー
 * @param lastState 最終ステート
 * @param attacker 攻撃側バッテリーコマンド
 * @param defender 防御側バッテリーコマンド
 * @return 更新されたゲームステート
 */
export function attackFlow(
  lastState: GameState,
  attacker: PlayerCommandX<BatteryCommand>,
  defender: PlayerCommandX<BatteryCommand>,
): GameState[] {
  return startGameFlow([
    (state) => [
      batteryDeclaration(
        state,
        attacker.playerId,
        attacker.command,
        defender.playerId,
        defender.command,
      ),
    ],
    (state) =>
      state.effect.name === "BatteryDeclaration"
        ? [
            battle(
              state,
              state.effect.attacker,
              state.effect.attackerBattery,
              defender.playerId,
              state.effect.defenderBattery,
            ),
          ]
        : [],
  ], lastState);
}
