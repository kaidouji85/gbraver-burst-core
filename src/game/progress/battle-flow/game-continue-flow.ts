import {
  canContinuousActive,
  continuousActive,
} from "../../../effect/continuous-active";
import { inputCommand } from "../../../effect/input-command";
import { turnChange } from "../../../effect/turn-change";
import { updateRemainingTurn } from "../../../effect/update-remaining-turn";
import type { GameState } from "../../../state/game-state";
import { startGameFlow } from "../../game-flow";

/**
 * ゲーム継続フロー
 * @param lastState 最新の状態
 * @returns 更新結果
 */
export function gameContinueFlow(lastState: GameState): GameState[] {
  return startGameFlow(lastState, [
    (state) => [updateRemainingTurn(state)],
    (state) =>
      canContinuousActive(state)
        ? [continuousActive(state)]
        : [turnChange(state)],
    (state) => [inputCommand({ lastState: state })],
  ]);
}
