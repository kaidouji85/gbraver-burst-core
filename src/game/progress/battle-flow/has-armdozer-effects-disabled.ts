import { PlayerState } from "../../../state/player-state";

/**
 * アームドーザ効果無効であるかを判定する
 * @param player 判定対象のプレイヤー
 * @returns 判定結果、trueでアームドーザ効果無効
 */
export const hasArmdozerEffectsDisabled = (player: PlayerState) =>
  player.armdozer.effects.some(
    (effect) => effect.type === "ArmdozerEffectsDisabled",
  );
