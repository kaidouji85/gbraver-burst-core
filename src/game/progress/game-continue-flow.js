// @flow

import type {BatteryCommand, Command, GameState, PlayerCommand, PlayerCommandX, PlayerId} from "../..";
import {gameFlow} from "../flow/game-flow";
import {updateRemainingTurn} from "../../effect/update-remaning-turn";
import {canContinuousActive, continuousActive} from "../../effect/continuous-active";
import {upcastGameState} from "../state/game-state";
import {turnChange} from "../../effect/turn-change";
import {inputCommand} from "../../effect/input-command";

/**
 * ゲーム継続フロー
 *
 * @param state 最新の状態
 * @param attackerId 攻撃側プレイヤーID
 * @param attackerCommand 攻撃側コマンド
 * @param defenderId 防御側プレイヤーID
 * @param defenderCommand 防御側コマンド
 * @return 更新結果
 */
export function gameContinueFlow(state: GameState, attackerId: PlayerId, attackerCommand: Command, defenderId: PlayerId, defenderCommand: Command): GameState[] {
  return gameFlow(state, [
    state => [updateRemainingTurn(state)],
    state => {
      if(canContinuousActive(state)) {
        const done = continuousActive(state);
        return done ? [upcastGameState(done)] : [];
      } else {
        const done = turnChange(state);
        return done ? [upcastGameState(done)] : [];
      }
    },
    state => {
      const done = inputCommand(
        state,
        attackerId,
        attackerCommand,
        defenderId,
        defenderCommand
      );
      return done ? [upcastGameState(done)] : [];
    }
  ]);
}