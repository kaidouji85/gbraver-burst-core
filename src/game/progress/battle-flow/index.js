// @flow
import type { BatteryCommand } from "../../../command/battery";
import type { Battle } from "../../../effect/battle/battle";
import { gameEnd } from "../../../effect/game-end";
import { canRightItself, rightItself } from "../../../effect/right-itself";
import type { GameState } from "../../../state/game-state";
import { upcastGameState } from "../../../state/game-state";
import type { PlayerCommandX } from "../../command/player-command";
import { gameEndJudging } from "../../end-judging";
import { startGameStateFlow } from "../../game-state-flow";
import { attackFlow } from "./attack-flow";
import { gameContinueFlow } from "./game-continue-flow";
import { canReflectFlow, reflectFlow } from "./reflect-flow";

/**
 * 戦闘フロー
 *
 * @param lastState 最後の状態
 * @param commands コマンド
 * @return 更新されたゲームステート
 */
export function battleFlow(
  lastState: GameState,
  commands: [PlayerCommandX<BatteryCommand>, PlayerCommandX<BatteryCommand>]
): GameState[] {
  const attacker = commands.find(
    (v) => v.playerId === lastState.activePlayerId
  );
  const defender = commands.find(
    (v) => v.playerId !== lastState.activePlayerId
  );
  if (!attacker || !defender) {
    throw new Error("not found attacker or defender command");
  }

  return startGameStateFlow(attackFlow(lastState, attacker, defender))
    .add((state) => {
      const battleEffect =
        state.effect.name === "Battle" ? (state.effect: Battle) : null;
      return battleEffect
        ? startGameStateFlow([state])
            .add((state) =>
              canReflectFlow(battleEffect.result)
                ? reflectFlow(state, attacker.playerId)
                : []
            )
            .add((state) =>
              canRightItself(battleEffect)
                ? [upcastGameState(rightItself(state, battleEffect))]
                : []
            )
            .toGameStateHistory()
            .slice(1)
        : [];
    })
    .add((state) => {
      const endJudge = gameEndJudging(state);
      return endJudge.type === "GameContinue"
        ? gameContinueFlow(
            state,
            attacker.playerId,
            attacker.command,
            defender.playerId,
            defender.command
          )
        : [upcastGameState(gameEnd(state, endJudge))];
    })
    .toGameStateHistory();
}
