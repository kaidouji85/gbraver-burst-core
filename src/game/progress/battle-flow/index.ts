import type { BatteryCommand } from "../../../command/battery";
import type { Battle } from "../../../effect/battle/battle";
import { gameEnd } from "../../../effect/game-end";
import { canRightItself, rightItself } from "../../../effect/right-itself";
import type { GameState } from "../../../state/game-state";
import type { PlayerCommandX } from "../../command/player-command";
import { gameEndJudging } from "../../end-judging";
import { startGameFlow } from "../../game-flow";
import { attackFlow } from "./attack-flow";
import { gameContinueFlow } from "./game-continue-flow";
import { canReflectFlow, reflectFlow } from "./reflect-flow";

/**
 * 戦闘フロー
 *
 * @param lastState 最後の状態
 * @param commands コマンド
 * @return 更新されたゲームステート
 */
export function battleFlow(
  lastState: GameState,
  commands: [PlayerCommandX<BatteryCommand>, PlayerCommandX<BatteryCommand>],
): GameState[] {
  const attacker = commands.find(
    (v) => v.playerId === lastState.activePlayerId,
  );
  const defender = commands.find(
    (v) => v.playerId !== lastState.activePlayerId,
  );
  if (!attacker || !defender) {
    throw new Error("not found attacker or defender command");
  }

  return startGameFlow(lastState, [
    (state) => attackFlow(state, attacker, defender),
    (state) => {
      if (state.effect.name === "Battle") {
        const battleEffect: Battle = state.effect;
        return startGameFlow(state, [
          (subState) =>
            canReflectFlow(battleEffect.result)
              ? reflectFlow(subState, attacker.playerId)
              : [],
          (subState) =>
            canRightItself(battleEffect)
              ? [rightItself(subState, battleEffect)]
              : [],
        ]);
      }
      return [];
    },
    (state) => {
      const endJudge = gameEndJudging(state);
      return endJudge.type === "GameContinue"
        ? gameContinueFlow(
            state,
            attacker.playerId,
            attacker.command,
            defender.playerId,
            defender.command,
          )
        : [gameEnd(state, endJudge)];
    },
  ]);
}
