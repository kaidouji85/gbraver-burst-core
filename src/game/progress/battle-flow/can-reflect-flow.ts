import { Battle } from "../../../effect/battle/battle";
import { isAttackHit } from "../../../effect/battle/result/is-attack-hit";
import { hasArmdozerEffectsDisabled } from "../../../effect/has-armdozer-effects-disabled";
import { PlayerState } from "../../../state/player-state";

/**
 * ダメージ反射フロー実施判定のオプション
 * 戦闘を行ったGameStateから必要なプロパティを抜き出したものである
 */
type Options = {
  /** 戦闘結果 */
  effect: Battle;
  /** 戦闘直後のプレイヤー情報 */
  players: PlayerState[];
};

/**
 * ダメージ反射フローを実施するか否かを判定する
 * @param options オプション
 */
export function canReflectFlow(options: Options) {
  const { effect, players } = options;
  const attacker = players.find((v) => v.playerId === effect.attacker);
  if (!attacker) {
    throw new Error("not found attacker");
  }

  return (
    !hasArmdozerEffectsDisabled(attacker.armdozer.effects) &&
    isAttackHit(effect.result)
  );
}
