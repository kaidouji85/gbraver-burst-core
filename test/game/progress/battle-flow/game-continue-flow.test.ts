import {
  BatteryCommand,
  EMPTY_ARMDOZER_STATE,
  EMPTY_GAME_STATE,
  EMPTY_PLAYER_STATE,
  GameState,
  PlayerState,
} from "../../../../src";
import { gameContinueFlow } from "../../../../src/game/progress/battle-flow/game-continue-flow";

test("ターン交代まで正しく実行できる", () => {
  const player1 = {
    ...EMPTY_PLAYER_STATE,
    playerId: "player1",
    armdozer: { ...EMPTY_ARMDOZER_STATE, maxBattery: 5, battery: 1 },
  };
  const player1Command: BatteryCommand = {
    type: "BATTERY_COMMAND",
    battery: 1,
  };
  const player2 = { ...EMPTY_PLAYER_STATE, playerId: "player2" };
  const player2Command: BatteryCommand = {
    type: "BATTERY_COMMAND",
    battery: 2,
  };
  const lastState = {
    ...EMPTY_GAME_STATE,
    players: [player1, player2],
    activePlayerId: player2.playerId,
  };
  const result = gameContinueFlow(
    lastState,
    player1.playerId,
    player1Command,
    player2.playerId,
    player2Command,
  );
  expect(result).toMatchSnapshot("turn-change");
});

test("アクティブプレイヤー継続を正しく処理できる", () => {
  const player1: PlayerState = {
    ...EMPTY_PLAYER_STATE,
    playerId: "player1",
    armdozer: {
      ...EMPTY_ARMDOZER_STATE,
      maxBattery: 5,
      battery: 3,
      effects: [
        {
          type: "ContinuousActivePlayer",
          period: {
            type: "SpecialPeriod",
          },
        },
      ],
    },
  };
  const player1Command: BatteryCommand = {
    type: "BATTERY_COMMAND",
    battery: 1,
  };
  const player2: PlayerState = { ...EMPTY_PLAYER_STATE, playerId: "player2" };
  const player2Command: BatteryCommand = {
    type: "BATTERY_COMMAND",
    battery: 2,
  };
  const lastState: GameState = {
    ...EMPTY_GAME_STATE,
    players: [player1, player2],
    activePlayerId: player1.playerId,
  };
  const result = gameContinueFlow(
    lastState,
    player1.playerId,
    player1Command,
    player2.playerId,
    player2Command,
  );
  expect(result).toMatchSnapshot("continuous-active");
});
