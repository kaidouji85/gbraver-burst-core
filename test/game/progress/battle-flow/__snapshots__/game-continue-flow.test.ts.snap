// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ゲーム継続フロー（アクティブプレイヤー継続）を正しく処理することができる: continuous-active 1`] = `
[
  {
    "activePlayerId": "player1",
    "effect": {
      "endArmdozerEffects": [],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "ContinuousActivePlayer",
            },
          ],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "TurnChange",
      "recoverBattery": 0,
    },
    "players": [
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player1",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player2",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 3,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
]
`;

exports[`ゲーム継続フロー（ターン交代）を正しく処理することができる: turn-change 1`] = `
[
  {
    "activePlayerId": "player2",
    "effect": {
      "endArmdozerEffects": [],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 1,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "TurnChange",
      "recoverBattery": 3,
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player1",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player1",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player2",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
]
`;
